### ISSUE TITLE
Automatic Mixed Precision with DataParallel

### ISSUE THREAD

**Issue by raphaelreme** [pytorch/pytorch#60540]
## 📚 Documentation

From what I've found as documentation for [AMP with DataParallel](https://pytorch.org/docs/stable/notes/amp_examples.html#dataparallel-in-a-single-process) Autocast should be done in all the threads created by DataParallel and not only on the main thread.

But I do not observe this behavior: Autocast seems to work in DataParallel without any effort from my side.

## Reproducing steps
Run the following code on a machine with several GPUs
```
import torch
import torch.nn as nn
from torch.utils.data import Dataset, DataLoader


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class Model(nn.Module):
    def __init__(self, input_size, output_size):
        super().__init__()
        self.fc = nn.Linear(input_size, output_size)

    def forward(self, input_):
        output = self.fc(input_)
        print(
            "\tInside: In size",
            input_.size(),
            "Out size",
            output.size(),
            "In device",
            input_.device,
            "Out device",
            output.device,
            "Autocast",
            torch.is_autocast_enabled(),
        )

        return output


def main():
    input_size = 5
    output_size = 2

    batch_size = 30
    data_size = 100
    autocast = True

    rand_loader = DataLoader(dataset=RandomDataset(input_size, data_size), batch_size=batch_size, shuffle=True)

    device = torch.device("cuda:0")
    model: nn.Module = Model(input_size, output_size).to(device)

    if torch.cuda.device_count() > 1:
        print("Let's use", torch.cuda.device_count(), "GPUs!")
        model = nn.DataParallel(model)

    for data in rand_loader:
        input_: torch.Tensor = data.to(device)

        with torch.cuda.amp.autocast(autocast):
            output: torch.Tensor = model(input_)
            print(
                "Outside: In size",
                input_.size(),
                "Out size",
                output.size(),
                "In Device",
                input_.device,
                "Out Device",
                output.device,
                "Autocast",
                torch.is_autocast_enabled(),
            )
        print("Out of autocast: is amp enabled ?", torch.is_autocast_enabled())
        print()


if __name__ == "__main__":
    main()
```

## Outputs on my computer

I've tried this code with both pytorch 1.8.1 and 1.9.0 and cudatoolkit 11.1
I have two GPUs, a RTX3090 and GTX1080TI

```
$ python test_AMP_DP.py 
Let's use 2 GPUs!
/home/raph/miniconda3/envs/navee/lib/python3.8/site-packages/torch/nn/parallel/data_parallel.py:30: UserWarning: 
    There is an imbalance between your GPUs. You may want to exclude GPU 1 which
    has less than 75% of the memory or cores of GPU 0. You can do so by setting
    the device_ids argument to DataParallel, or by setting the CUDA_VISIBLE_DEVICES
    environment variable.
  warnings.warn(imbalance_warn.format(device_ids[min_pos], device_ids[max_pos]))
	Inside: In size torch.Size([15, 5]) Out size torch.Size([15, 2]) In device cuda:1 Out device cuda:1 Autocast True
	Inside: In size torch.Size([15, 5]) Out size torch.Size([15, 2]) In device cuda:0 Out device cuda:0 Autocast True
Outside: In size torch.Size([30, 5]) Out size torch.Size([30, 2]) In Device cuda:0 Out Device cuda:0 Autocast True
Out of autocast: is amp enabled ? False

	Inside: In size torch.Size([15, 5]) Out size torch.Size([15, 2]) In device cuda:0 Out device cuda:0 Autocast True
	Inside: In size torch.Size([15, 5]) Out size torch.Size([15, 2]) In device cuda:1 Out device cuda:1 Autocast True
Outside: In size torch.Size([30, 5]) Out size torch.Size([30, 2]) In Device cuda:0 Out Device cuda:0 Autocast True
Out of autocast: is amp enabled ? False

	Inside: In size torch.Size([15, 5]) Out size torch.Size([15, 2]) In device cuda:0 Out device cuda:0 Autocast True
	Inside: In size torch.Size([15, 5]) Out size torch.Size([15, 2]) In device cuda:1 Out device cuda:1 Autocast True
Outside: In size torch.Size([30, 5]) Out size torch.Size([30, 2]) In Device cuda:0 Out Device cuda:0 Autocast True
Out of autocast: is amp enabled ? False

	Inside: In size torch.Size([5, 5]) Out size torch.Size([5, 2]) In device cuda:0 Out device cuda:0 Autocast True
	Inside: In size torch.Size([5, 5]) Out size torch.Size([5, 2]) In device cuda:1 Out device cuda:1 Autocast True
Outside: In size torch.Size([10, 5]) Out size torch.Size([10, 2]) In Device cuda:0 Out Device cuda:0 Autocast True
Out of autocast: is amp enabled ? False
```

As one can see AMP is enabled inside the forward pass even though I only activated it from the main thread. (It's a good news for me, but that's not what the documentation says)

(The warning is expected as the second GPU is much smaller than the first one. And we can see that DP works as the inputs are scattered on the two GPUs)

I"m also observing this behavior on the same environment with real life code/training loop.

## Expected outputs
```
$ python test_AMP_DP.py 
Let's use 2 GPUs!
/home/raph/miniconda3/envs/navee/lib/python3.8/site-packages/torch/nn/parallel/data_parallel.py:30: UserWarning: 
    There is an imbalance between your GPUs. You may want to exclude GPU 1 which
    has less than 75% of the memory or cores of GPU 0. You can do so by setting
    the device_ids argument to DataParallel, or by setting the CUDA_VISIBLE_DEVICES
    environment variable.
  warnings.warn(imbalance_warn.format(device_ids[min_pos], device_ids[max_pos]))
	Inside: In size torch.Size([15, 5]) Out size torch.Size([15, 2]) In device cuda:1 Out device cuda:1 Autocast False
	Inside: In size torch.Size([15, 5]) Out size torch.Size([15, 2]) In device cuda:0 Out device cuda:0 Autocast False
Outside: In size torch.Size([30, 5]) Out size torch.Size([30, 2]) In Device cuda:0 Out Device cuda:0 Autocast True
Out of autocast: is amp enabled ? False

	Inside: In size torch.Size([15, 5]) Out size torch.Size([15, 2]) In device cuda:0 Out device cuda:0 Autocast False
	Inside: In size torch.Size([15, 5]) Out size torch.Size([15, 2]) In device cuda:1 Out device cuda:1 Autocast False
Outside: In size torch.Size([30, 5]) Out size torch.Size([30, 2]) In Device cuda:0 Out Device cuda:0 Autocast True
Out of autocast: is amp enabled ? False

	Inside: In size torch.Size([15, 5]) Out size torch.Size([15, 2]) In device cuda:0 Out device cuda:0 Autocast False
	Inside: In size torch.Size([15, 5]) Out size torch.Size([15, 2]) In device cuda:1 Out device cuda:1 Autocast False
Outside: In size torch.Size([30, 5]) Out size torch.Size([30, 2]) In Device cuda:0 Out Device cuda:0 Autocast True
Out of autocast: is amp enabled ? False

	Inside: In size torch.Size([5, 5]) Out size torch.Size([5, 2]) In device cuda:0 Out device cuda:0 Autocast False
	Inside: In size torch.Size([5, 5]) Out size torch.Size([5, 2]) In device cuda:1 Out device cuda:1 Autocast False
Outside: In size torch.Size([10, 5]) Out size torch.Size([10, 2]) In Device cuda:0 Out Device cuda:0 Autocast True
Out of autocast: is amp enabled ? False
```

-------------------------------------------------------------------------------------------------

Any ideas on what I'm missing here ? And what is the expected behavior, the one of the doc or the one I'm seeing ?

@mcarilli 

cc @mcarilli @ptrblck

**Comment by raphaelreme** [comment]
Any news on this subject ?

**Comment by RameshArvind** [comment]
I recently was experiencing this issue and digging around led me to this old [PR](https://github.com/pytorch/pytorch/pull/43102#issuecomment-679244117)
TL;DR The documentation is outdated. `autocast` works with `DataParallel` seamlessly

**Comment by raphaelreme** [comment]
Ah great ! Exactly what I was looking for thank you !
Let's close this issue :)

(Found the merged commit: f02753fabb8116ab202fd56b3fcaf9389b7cade6)
